### PR TITLE
feat: support neon, sse simd and dynamic dispatch

### PR DESCRIPTION
**Pull request by xiegx94**
# Main changes
- support both static dispatch and dynamic dispatch
- support neon and sse architecture

### CODE REVIEW

**Comment by xiegx94** [comment]
This PR provides 2 ways to support multi-arch dispatch: dispatch at compile (static dispatch) and dispatch at runtime (dynamic dispatch).  Dynamic dispatch is implemented by using gcc/clang multiversioning-functions, which causes these function cannot be inlined when compile and the performance will be worse.

# The structure of arch folder
```bash
├── avx2
│   ├── base.h
│   ├── itoa.h
│   ├── quote.h
│   ├── simd.h
│   ├── skip.h
│   ├── str2int.h
│   └── unicode.h
├── common
│   ├── quote_common.h
│   ├── quote_tables.h
│   ├── skip_common.h
│   ├── unicode_common.h
│   └── x86_common
│       ├── itoa.h
│       ├── quote.inc.h
│       └── skip.inc.h
├── neon
│   ├── base.h
│   ├── itoa.h
│   ├── quote.h
│   ├── simd.h
│   ├── skip.h
│   ├── str2int.h
│   └── unicode.h
├── simd_base.h
├── simd_dispatch.h
├── simd_itoa.h
├── simd_quote.h
├── simd_skip.h
├── simd_str2int.h
├── sonic_cpu_feature.h
├── sse
│   ├── base.h
│   ├── itoa.h
│   ├── quote.h
│   ├── simd.h
│   ├── skip.h
│   ├── str2int.h
│   └── unicode.h
├── target_macro.h
└── x86_ifuncs
    ├── base.h
    ├── ifunc_macro.h
    ├── itoa.h
    ├── quote.h
    ├── skip.h
    └── str2int.h
```

# How to add a new function
If you want to add a new simd function which is called `foo`, then, you should follow below steps:
1. implement foo for every arch, such as:
```c++
namespace sonic_json {
namespace internal {
namespace avx2 {

void foo() { return; }

}  // namespace avx2
}  // namespace internal
}  // namespace sonic_json
```
2. provide dynamic dispatch functions for x86 (or others platform)
```c++
namespace sonic_json {
namespace internal {

__attribute__((target(HASWELL))) inline void foo() { return avx2::foo(); }
__attribute__((target(WESTMERE))) inline void foo() { return sse::foo(); }
__attribute__((target("default"))) inline void foo() { return sse::foo(); }

}
}
```
3. If you want implement `foo` in a new header file `foo.h`, you should provide such file for every arch and x86_ifuncs. then add a new file `simd_foo.h` in arch floder:
```c++
#pragma once

#include "simd_dispatch.h"

#include INCLUDE_ARCH_FILE(foo.h)

namespace sonic_json {
namespace internal {

SONIC_USING_ARCH_FUNC(foo);

}
}
``` 
# How to add a new architecture
If there is a new architecture named `Y86`, you should do:
1. write a new rule to detect `Y86` macro ( provide by gcc/clang) in `sonic_cpu_feature.h`
```c++
#if defined(__Y86__)
#define SONIC_HAVE_Y86
#endif
```
2. write a new rule about how to dispatch in `simd_dispatch.h`
```c++
#if defined(SONIC_STATIC_DISPATCH)
#if defined(SONIC_HAVE_Y86)
#define SONIC_USING_ARCH_FUNC(func) using Y86::func
#define INCLUDE_ARCH_FILE(file) SONIC_STRINGIFY(Y86/file)
#endif
#elif defined(SONIC_DYNAMIC_DISPATCH)
#if defined(SONIC_HAVE_Y86)
#define SONIC_USING_ARCH_FUNC(func)
#define INCLUDE_ARCH_FILE(file) SONIC_STRINGIFY(y86_ifuncs/file)
#endif
#endif
```
3. create y86 folder and implement all simd functions
4. create y86 folder and implement all multiversioning-functions.

----
sonic 的多架构设计同时支持在编译期间选择指定的指令和在运行时根据运行的平台选择合适的指令。同时支持两种方式是因为在运行时抉择会让使用 simd 的函数/接口无法在编译期间 inline，这会引起一些性能下降。
# arch 目录结构
```bash 
├── avx2
│   ├── base.h
│   ├── itoa.h
│   ├── quote.h
│   ├── simd.h
│   ├── skip.h
│   ├── str2int.h
│   └── unicode.h
├── common
│   ├── quote_common.h
│   ├── quote_tables.h
│   ├── skip_common.h
│   ├── unicode_common.h
│   └── x86_common
│       ├── itoa.h
│       ├── quote.inc.h
│       └── skip.inc.h
├── neon
│   ├── base.h
│   ├── itoa.h
│   ├── quote.h
│   ├── simd.h
│   ├── skip.h
│   ├── str2int.h
│   └── unicode.h
├── simd_base.h
├── simd_dispatch.h
├── simd_itoa.h
├── simd_quote.h
├── simd_skip.h
├── simd_str2int.h
├── sonic_cpu_feature.h
├── sse
│   ├── base.h
│   ├── itoa.h
│   ├── quote.h
│   ├── simd.h
│   ├── skip.h
│   ├── str2int.h
│   └── unicode.h
├── target_macro.h
└── x86_ifuncs
    ├── base.h
    ├── ifunc_macro.h
    ├── itoa.h
    ├── quote.h
    ├── skip.h
    └── str2int.h
```
avx2, sse, neon。特定架构下的 simd 实现代码
common, 通用的一些实现
x86_ifuncs x86 平台动态 dispatch 代码
# 如何添加新的函数
1. 在每个 arch 下添加新的函数，如：
```c++
namespace sonic_json {
namespace internal {
namespace avx2 {

void foo() { return; }

}  // namespace avx2
}  // namespace internal
}  // namespace sonic_json
```
5. 在 x86_ifunc 下添加 x86 动态 dispatch 支持：
```c++
namespace sonic_json {
namespace internal {

__attribute__((target(HASWELL))) inline void foo() { return avx2::foo(); }
__attribute__((target(WESTMERE))) inline void foo() { return sse::foo(); }
__attribute__((target("default"))) inline void foo() { return sse::foo(); }

}
}
```
6. （可选）如果添加了新的头文件，则在 arch 下添加 simd_foo.h，在各 arch 下添加 foo.h 文件。simd_foo 如下：
```c++
#pragma once

#include "simd_dispatch.h"

#include INCLUDE_ARCH_FILE(foo.h)

namespace sonic_json {
namespace internal {

SONIC_USING_ARCH_FUNC(foo);

}
}
```
# 如何添加新的架构
假如有个新的架构叫Y86，需要在 sonic 中添加其 simd 支持，则：
1. 在 sonic_cpu_feature.h 中检测Y86的宏:
```c++
#if defined(__Y86__)
#define SONIC_HAVE_Y86
#endif
```
2. 在 simd_dispatch 中添加 dispatch 规则
```c++
#if defined(SONIC_STATIC_DISPATCH)
#if defined(SONIC_HAVE_Y86)
#define SONIC_USING_ARCH_FUNC(func) using Y86::func
#define INCLUDE_ARCH_FILE(file) SONIC_STRINGIFY(Y86/file)
#endif
#elif defined(SONIC_DYNAMIC_DISPATCH)
#if defined(SONIC_HAVE_Y86)
#define SONIC_USING_ARCH_FUNC(func)
#define INCLUDE_ARCH_FILE(file) SONIC_STRINGIFY(y86_ifuncs/file)
#endif
#endif
```
3. 添加 y86 文件夹，添加所有的 simd 函数的 y86 实现
4. 添加 y86_ifuncs 文件夹，添加 y86 的 multiversioning-function 实现

**Comment by xiegx94** [comment]
**Performance**
|test case                      |   sse        |   haswell    |  dynamic dispatch |
|-------------------------------|--------------|--------------|-------------------|
|book/Decode_SonicDyn           |       813 ns |       849 ns |      1128 ns |
|gsoc-2018/Decode_SonicDyn      |   1470898 ns |   1339296 ns |   1752588 ns |
|fgo/Decode_SonicDyn            | 112165070 ns | 117719364 ns | 150338769 ns |
|lottie/Decode_SonicDyn         |    805143 ns |    842756 ns |   1187414 ns |
|canada/Decode_SonicDyn         |   3756878 ns |   3789520 ns |   4085432 ns |
|github_events/Decode_SonicDyn  |     39368 ns |     39716 ns |     54755 ns |
|otfcc/Decode_SonicDyn          | 292141676 ns | 320360184 ns | 377427578 ns |
|poet/Decode_SonicDyn           |   1572923 ns |   1534339 ns |   1743444 ns |
|citm_catalog/Decode_SonicDyn   |   1137476 ns |   1217241 ns |   1439325 ns |
|twitter/Decode_SonicDyn        |    181451 ns |    185673 ns |    260165 ns |
|twitterescaped/Decode_SonicDyn |    492546 ns |    555098 ns |    671564 ns |
|book/Encode_SonicDyn           |       619 ns |       631 ns |       616 ns |
|gsoc-2018/Encode_SonicDyn      |    796307 ns |    680203 ns |    672574 ns |
|fgo/Encode_SonicDyn            |  75432301 ns |  76930374 ns |  75568517 ns |
|lottie/Encode_SonicDyn         |    858753 ns |    839591 ns |    871623 ns |
|canada/Encode_SonicDyn         |   6152427 ns |   6009102 ns |   6074922 ns |
|github_events/Encode_SonicDyn  |     22432 ns |     21035 ns |     20380 ns |
|otfcc/Encode_SonicDyn          | 155490041 ns | 159626245 ns | 160975575 ns |
|poet/Encode_SonicDyn           |    840780 ns |    720925 ns |    721330 ns |
|citm_catalog/Encode_SonicDyn   |    533967 ns |    560394 ns |    554867 ns |
|twitter/Encode_SonicDyn        |     97690 ns |     93807 ns |     89528 ns |
|twitterescaped/Encode_SonicDyn |    284284 ns |    263186 ns |    269042 ns |

File: include/sonic/internal/arch/avx2/simd.h [hunk 1]
```diff
@@ -479,7 +515,15 @@ struct simd256<int8_t> : num256<int8_t> {
 template <>
 struct simd256<uint8_t> : num256<uint8_t> {
   using Base = num256<uint8_t>;
-  using Base::Base;
+  // using Base::Base;
```
Comment (liuq19) [inline include/sonic/internal/arch/avx2/simd.h:518]: 这里为啥不能复用，按理说O3 优化下，应该都能内联

**Comment by liuq19** [comment]
> **Performance**
> 
> test case	sse	haswell	dynamic dispatch
> book/Decode_SonicDyn	813 ns	849 ns	1128 ns
> gsoc-2018/Decode_SonicDyn	1470898 ns	1339296 ns	1752588 ns
> fgo/Decode_SonicDyn	112165070 ns	117719364 ns	150338769 ns
> lottie/Decode_SonicDyn	805143 ns	842756 ns	1187414 ns
> canada/Decode_SonicDyn	3756878 ns	3789520 ns	4085432 ns
> github_events/Decode_SonicDyn	39368 ns	39716 ns	54755 ns
> otfcc/Decode_SonicDyn	292141676 ns	320360184 ns	377427578 ns
> poet/Decode_SonicDyn	1572923 ns	1534339 ns	1743444 ns
> citm_catalog/Decode_SonicDyn	1137476 ns	1217241 ns	1439325 ns
> twitter/Decode_SonicDyn	181451 ns	185673 ns	260165 ns
> twitterescaped/Decode_SonicDyn	492546 ns	555098 ns	671564 ns
> book/Encode_SonicDyn	619 ns	631 ns	616 ns
> gsoc-2018/Encode_SonicDyn	796307 ns	680203 ns	672574 ns
> fgo/Encode_SonicDyn	75432301 ns	76930374 ns	75568517 ns
> lottie/Encode_SonicDyn	858753 ns	839591 ns	871623 ns
> canada/Encode_SonicDyn	6152427 ns	6009102 ns	6074922 ns
> github_events/Encode_SonicDyn	22432 ns	21035 ns	20380 ns
> otfcc/Encode_SonicDyn	155490041 ns	159626245 ns	160975575 ns
> poet/Encode_SonicDyn	840780 ns	720925 ns	721330 ns
> citm_catalog/Encode_SonicDyn	533967 ns	560394 ns	554867 ns
> twitter/Encode_SonicDyn	97690 ns	93807 ns	89528 ns
> twitterescaped/Encode_SonicDyn	284284 ns	263186 ns	269042 ns

最好分别贴下static 模式和 dynamic 模式下，目前分支和master分支的相对性能测试数据，这样应该更清楚一点

File: include/sonic/internal/arch/avx2/simd.h (hunk 1 above)
Comment (xiegx94) [reply include/sonic/internal/arch/avx2/simd.h:518]: 和内联没关系。这里没有提供构造函数，所以构造函数是默认的，没有内联。这样会导致构造函数在非指定的架构下被应用，引发编译器报错。

File: include/sonic/internal/arch/x86_ifuncs/skip.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+#include "../avx2/skip.h"
+#include "../sse/skip.h"
+#include "ifunc_macro.h"
+
+namespace sonic_json {
+namespace internal {
+
+using common::EqBytes4;
+using common::SkipLiteral;
+using sse::GetNextToken;  // !!!Not efficency
```
[the remark this answers is not in the document]
Comment (xiegx94) [reply include/sonic/internal/arch/x86_ifuncs/skip.h:28]: GetNextToken 是一个模板函数，没办法使用 multiversion function 的机制让编译器自动选择版本。这里在所有的架构下都选择了sse 版本。

**Comment by xiegx94** [comment]
> > **Performance**
> > test case	sse	haswell	dynamic dispatch
> > book/Decode_SonicDyn	813 ns	849 ns	1128 ns
> > gsoc-2018/Decode_SonicDyn	1470898 ns	1339296 ns	1752588 ns
> > fgo/Decode_SonicDyn	112165070 ns	117719364 ns	150338769 ns
> > lottie/Decode_SonicDyn	805143 ns	842756 ns	1187414 ns
> > canada/Decode_SonicDyn	3756878 ns	3789520 ns	4085432 ns
> > github_events/Decode_SonicDyn	39368 ns	39716 ns	54755 ns
> > otfcc/Decode_SonicDyn	292141676 ns	320360184 ns	377427578 ns
> > poet/Decode_SonicDyn	1572923 ns	1534339 ns	1743444 ns
> > citm_catalog/Decode_SonicDyn	1137476 ns	1217241 ns	1439325 ns
> > twitter/Decode_SonicDyn	181451 ns	185673 ns	260165 ns
> > twitterescaped/Decode_SonicDyn	492546 ns	555098 ns	671564 ns
> > book/Encode_SonicDyn	619 ns	631 ns	616 ns
> > gsoc-2018/Encode_SonicDyn	796307 ns	680203 ns	672574 ns
> > fgo/Encode_SonicDyn	75432301 ns	76930374 ns	75568517 ns
> > lottie/Encode_SonicDyn	858753 ns	839591 ns	871623 ns
> > canada/Encode_SonicDyn	6152427 ns	6009102 ns	6074922 ns
> > github_events/Encode_SonicDyn	22432 ns	21035 ns	20380 ns
> > otfcc/Encode_SonicDyn	155490041 ns	159626245 ns	160975575 ns
> > poet/Encode_SonicDyn	840780 ns	720925 ns	721330 ns
> > citm_catalog/Encode_SonicDyn	533967 ns	560394 ns	554867 ns
> > twitter/Encode_SonicDyn	97690 ns	93807 ns	89528 ns
> > twitterescaped/Encode_SonicDyn	284284 ns	263186 ns	269042 ns
> 
> 最好分别贴下static 模式和 dynamic 模式下，目前分支和master分支的相对性能测试数据，这样应该更清楚一点

Updated.

File: include/sonic/internal/arch/avx2/quote.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+#include "../common/quote_common.h"
+#include "../common/quote_tables.h"
+#include "base.h"
+#include "simd.h"
+#include "unicode.h"
+
+#if defined(__clang__)
+#pragma clang attribute push(__attribute__((target("avx2"))), \
+                             apply_to = function)
+#elif defined(__GNUG__)
```
Comment (liuq19) [inline include/sonic/internal/arch/avx2/quote.h:34]: 这里可以统一用一个宏概括

Comment (xiegx94) [reply include/sonic/internal/arch/avx2/quote.h:34]: How?

File: BUILD.bazel [hunk 4]
```diff
@@ -34,12 +99,21 @@ cc_binary(
...
-    copts = common_copts + ['-DNDEBUG', '-std=c++17'],
+    # copts = common_copts + ['-DNDEBUG', '-std=c++17'],
+    copts = select({
+                "arm_build": arm_copts + benchmark_copts,
+                "sse_build": sse_copts + benchmark_copts,
+                "avx2_build": avx2_copts + benchmark_copts,
+            }) +\
+            select({
+                "static_dispatch": static_dispatch_copts,
+                "dynamic_dispatch": dynamic_dispatch_copts,
```
[the remark this answers is not in the document]
Comment (xiegx94) [reply BUILD.bazel:110]: 这里没有指定必须使用 mavx2 编译选项。

File: include/sonic/allocator.h [hunk 5]
```diff
@@ -89,7 +89,12 @@ class SpinLock {
         break;
       }
       while (lock_.load(std::memory_order_relaxed)) {
+#if defined(__x86_64__) || defined(_M_AMD64)
         __builtin_ia32_pause();
+#elif defined(__aarch64__) || defined(_M_ARM64)
+        asm volatile("yield");
+#else
```
[the remark this answers is not in the document]
Comment (xiegx94) [reply include/sonic/allocator.h:96]: ok

File: include/sonic/internal/arch/x86_ifuncs/skip.h [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+__attribute__((target("default"))) inline uint8_t skip_space(const uint8_t*,
+                                                             size_t&, size_t&,
+                                                             uint64_t&) {
+  return 0;
+}
+
+__attribute__((target("default"))) inline uint8_t skip_space_safe(
+    const uint8_t*, size_t&, size_t, size_t&, uint64_t&) {
+  return 0;
```
[the remark this answers is not in the document]
Comment (xiegx94) [reply include/sonic/internal/arch/x86_ifuncs/skip.h:49]: 是的，在没有实现 fallback 的时候，这里使用 static assert 阻止编译比较好。

File: include/sonic/internal/arch/neon/str2int.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+namespace neon {
+
+sonic_force_inline uint64_t simd_str2int(const char* c, int& man_nd) {
+  uint64_t sum = 0;
+  int i = 0;
+  while (c[i] >= '0' && c[i] <= '9' && i < man_nd) {
+    sum = sum * 10 + (c[i] - '0');
+    i++;
+  }
+  man_nd = i;
```
[the remark this answers is not in the document]
Comment (xiegx94) [reply include/sonic/internal/arch/neon/str2int.h:32]: 暂时没想到在 neon 下怎么实现这个函数

**Pull request merged**